src/base/bittorrent/session.cpp

```diff
@@ -2645,6 +2654,62 @@ void Session::setAltGlobalUploadSpeedLimit(int limit)
         configureDeferred();
 }
 
+int Session::altPauseDownloads() const
+{
+    return m_isAltPauseDownloadsEnabled;
+}
+
+void Session::setAltPauseDownloads(bool enabled)
+{
+    m_isAltPauseDownloadsEnabled = enabled;
+}
+
+int Session::altPauseUploads() const
+{
+    return m_isAltPauseUploadsEnabled;
+}
+
+void Session::setAltPauseUploads(bool enabled)
+{
+    m_isAltPauseUploadsEnabled = enabled;
+}
+
+void Session::applyAltPauseDownloads(bool enabled)
+{
+    if (enabled) {
+        foreach (TorrentHandle *const torrent, m_torrents) {
+            if (torrent->isDownloading()) {
+                torrent->pause();
+            }
+        }
+    }
+    else {
+        foreach (TorrentHandle *const torrent, m_torrents) {
+            //torrent is paused and not completed, resume download
+            if (torrent->isPaused() && !torrent->isCompleted())
+                torrent->resume();
+        }
+    }
+}
+
+void Session::applyAltPauseUploads(bool enabled)
+{
+    if (enabled) {
+        foreach (TorrentHandle *const torrent, m_torrents) {
+            if (torrent->isUploading()) {
+                torrent->pause();
+            }
+        }
+    }
+    else {
+        foreach (TorrentHandle *const torrent, m_torrents) {
+            //torrent is paused and completed, resume upload
+            if (torrent->isPaused() && torrent->isCompleted())
```

Comment:
I wonder if it's worth it to check if the torrent has already reached their ratio/seeding time limit and prevent it from resuming in the first place (it'll be re-paused soon after anyway). Here's some code for it (untested), mostly copied from `Session::processShareLimits()`

```
if (torrent->ratioLimit() != TorrentHandle::NO_RATIO_LIMIT) {
    const qreal ratio = torrent->realRatio();
    qreal ratioLimit = torrent->ratioLimit();
    if (ratioLimit == TorrentHandle::USE_GLOBAL_RATIO)
         ratioLimit = globalMaxRatio();
    if (ratioLimit >= 0)
        if ((ratio <= TorrentHandle::MAX_RATIO) && (ratio >= ratioLimit))
            continue;
}
if (torrent->seedingTimeLimit() != TorrentHandle::NO_SEEDING_TIME_LIMIT) {
    const int seedingTimeInMinutes = torrent->seedingTime() / 60;
    int seedingTimeLimit = torrent->seedingTimeLimit();
    if (seedingTimeLimit == TorrentHandle::USE_GLOBAL_SEEDING_TIME)
        seedingTimeLimit = globalMaxSeedingMinutes();
    if (seedingTimeLimit >= 0)
        if ((seedingTimeInMinutes <= TorrentHandle::MAX_SEEDING_TIME) && (seedingTimeInMinutes >= seedingTimeLimit))
            continue;
}

Reply:
I don't know about code structure, do what you think is best or wait for an answer from a real dev :)